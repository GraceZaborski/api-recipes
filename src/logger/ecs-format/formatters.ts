// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { formatError } from '@elastic/ecs-helpers';

interface LogObject {
  error: Record<string, unknown>;
  ecs: { version: string };
  'event.duration': number;
}

const NANOSECONDS = 1000000;

function bindings(bindingsObject: Record<string, unknown>) {
  const { pid, hostname, ...ecsBindings } = bindingsObject;

  if (pid !== undefined) {
    ecsBindings.process = { pid };
  }

  if (hostname !== undefined) {
    ecsBindings.host = { hostname };
  }

  return ecsBindings;
}

function level(label: string) {
  return { 'log.level': label };
}

function log(object: LogObject) {
  const { error, ...ecsObject } = object;
  ecsObject.ecs = { version: '1.6.0' };

  if (ecsObject['event.duration'] !== undefined) {
    // event.duration is in milliseconds
    ecsObject['event.duration'] *= NANOSECONDS;
  }

  // mutates
  formatError(ecsObject, error);

  return ecsObject;
}

export const formatters = {
  bindings,
  level,
  // eslint-disable-next-line @typescript-eslint/ban-types
  log: (o: object) => log(o as LogObject),
};
