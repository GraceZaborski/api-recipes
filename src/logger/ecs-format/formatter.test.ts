import { formatters } from './formatters';

describe('level', () => {
  const { level } = formatters;

  it('should format log level', () => {
    expect(level('mock-level')).toEqual({
      'log.level': 'mock-level',
    });
  });
});

describe('bindings', () => {
  const { bindings } = formatters;

  it('should format bindings', () => {
    expect(
      bindings({
        pid: 'mock-pid',
        hostname: 'mock-hostname',
      }),
    ).toMatchSnapshot();
  });

  it('should ignore missing fields', () => {
    expect(bindings({})).toEqual({});
  });
});

describe('log', () => {
  const { log } = formatters;

  it('should format error', () => {
    const error = new Error('mock-error');
    error.stack = 'mock-stack-trace';

    const req = {
      key: 'mock-key',
      error,
    };

    expect(log(req)).toMatchSnapshot();
  });

  it('should identify type error', () => {
    const error = new TypeError('mock-error');
    error.stack = 'mock-stack-trace';
    const req = {
      key: 'mock-key',
      error,
    };

    expect(log(req)).toMatchSnapshot();
  });

  it('should pass through error object unchanged', () => {
    const req = {
      error: { foo: 'bar' },
      key: 'mock-key',
    };

    expect(log(req)).toMatchSnapshot();
  });

  it('should format the event.duration to nanoseconds', () => {
    const req = {
      error: { foo: 'bar' },
      key: 'mock-key',
      'event.duration': 192,
    };

    expect(log(req)).toMatchSnapshot();
  });
});
