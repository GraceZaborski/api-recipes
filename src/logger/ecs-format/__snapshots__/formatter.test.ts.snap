// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bindings should format bindings 1`] = `
Object {
  "host": Object {
    "hostname": "mock-hostname",
  },
  "process": Object {
    "pid": "mock-pid",
  },
}
`;

exports[`log should format error 1`] = `
Object {
  "ecs": Object {
    "version": "1.6.0",
  },
  "error": Object {
    "message": "mock-error",
    "stack_trace": "mock-stack-trace",
    "type": "Error",
  },
  "key": "mock-key",
}
`;

exports[`log should format the event.duration to nanoseconds 1`] = `
Object {
  "ecs": Object {
    "version": "1.6.0",
  },
  "err": Object {
    "foo": "bar",
  },
  "event.duration": 192000000,
  "key": "mock-key",
}
`;

exports[`log should identify type error 1`] = `
Object {
  "ecs": Object {
    "version": "1.6.0",
  },
  "error": Object {
    "message": "mock-error",
    "stack_trace": "mock-stack-trace",
    "type": "TypeError",
  },
  "key": "mock-key",
}
`;

exports[`log should pass through error object unchanged 1`] = `
Object {
  "ecs": Object {
    "version": "1.6.0",
  },
  "err": Object {
    "foo": "bar",
  },
  "key": "mock-key",
}
`;
